cf-oscore/src/main/java/org/eclipse/californium/oscore/ObjectSecurityLayer.java

```diff
@@ -152,6 +152,16 @@ public void sendRequest(Exchange exchange, Request request) {
 					@Override
 					public void onReadyToSend() {
 						Token token = preparedRequest.getToken();
+
+						// For Block-Wise requests update the Token in the
+						// original request so the request/response matching
+						// works. (OSCORE creates a new request so the
+						// MessageObserver attached in the BlockwiseLayer will
+						// not do this.)
+						if (request.getOptions().hasBlock1()) {
```

Comment:
The blockwise message observers are just try to read that from the "block-request" and transfer it to the "outer-request". But the "block-request", is replaced by the "prepared-request" and therefore don't contain that token any longer. To fix that, I introduced right now (master) a `Message.add(index, messageObserver)` and if that MessageObserver is added at the head, it works. It transfers the token from the "prepared-request" to the "block-request" before the other observer then forwards this token again.

```
preparedRequest.addMessageObserver(0, new MessageObserverAdapter() {

	@Override
	public void onReadyToSend() {
		Token token = preparedRequest.getToken();

		// add at had of message observers to update
		// the token of the original request first,
		// before calling other message observers!
		if (request.getToken() == null) {
			request.setToken(token);
		}

		ctxDb.addContext(token, finalCtx);
		ctxDb.addSeqByToken(token, seqByToken);
	}
});
```

Reply:
Thanks alot. I will update the code to use this method.

Reply:
I have now updated the code to use this method instead in commit b470ba265e78cf61c898039623e0fa4586e48614